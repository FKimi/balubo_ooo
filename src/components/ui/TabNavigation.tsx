'use client'

import React from 'react'

export interface TabConfig {
  key: string
  label: string
  count?: number
  icon?: React.ReactNode
}

interface TabNavigationProps {
  tabs: TabConfig[]
  activeTab: string
  onTabChange: (_tab: string) => void
  className?: string
}

export function TabNavigation({ tabs, activeTab, onTabChange, className = '' }: TabNavigationProps) {
  return (
    <div className={`flex justify-center ${className}`}>
      <div className="flex items-center gap-1 p-1 bg-gray-50/30 rounded-2xl">
        {tabs.map((tab) => (
        <button
          key={tab.key}
          onClick={() => onTabChange(tab.key)}
          className={`relative flex items-center gap-2 px-6 py-3 text-sm font-medium transition-all duration-300 ease-out ${
            activeTab === tab.key
              ? 'text-[#1e3a8a]'
              : 'text-gray-500 hover:text-gray-700'
          }`}
        >
          {/* アイコン */}
          {tab.icon && (
            <div className={`w-4 h-4 transition-colors duration-300 ${
              activeTab === tab.key ? 'text-[#1e3a8a]' : 'text-gray-400'
            }`}>
              {tab.icon}
            </div>
          )}
          
          {/* ラベル */}
          <span className="whitespace-nowrap">{tab.label}</span>
          
          {/* カウント */}
          {tab.count !== undefined && (
            <span className={`text-xs font-normal ${
              activeTab === tab.key
                ? 'text-[#1e3a8a]/70'
                : 'text-gray-400'
            }`}>
              {tab.count}
            </span>
          )}
          
          {/* アクティブ時の背景 */}
          {activeTab === tab.key && (
            <div className="absolute inset-0 bg-white rounded-xl shadow-sm border border-gray-100 -z-10"></div>
          )}
        </button>
        ))}
      </div>
    </div>
  )
} 