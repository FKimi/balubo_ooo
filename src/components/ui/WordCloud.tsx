'use client'

import React from 'react'
import { TagCloud } from 'react-tagcloud'

interface WordCloudData {
  value: string
  count: number
}

interface WordCloudProps {
  data: WordCloudData[]
  minSize?: number
  maxSize?: number
  className?: string
}

export function WordCloud({ 
  data, 
  minSize = 12, 
  maxSize = 35, 
  className = '' 
}: WordCloudProps) {
  // データが空の場合は何も表示しない
  if (!data || data.length === 0) {
    return (
      <div className={`flex items-center justify-center h-32 text-gray-400 ${className}`}>
        <p className="text-sm">データがありません</p>
      </div>
    )
  }

  // react-tagcloudの形式に変換
  const tagCloudData = data.map(item => ({
    value: item.value,
    count: item.count
  }))

  return (
    <div className={`word-cloud-container ${className}`}>
      <TagCloud
        minSize={minSize}
        maxSize={maxSize}
        tags={tagCloudData}
        className="word-cloud"
        onClick={(_tag: any) => {
          // タグクリック時の処理（必要に応じて実装）
        }}
        renderer={(_tag: any, size: number, color: string) => (
          <span
            key={_tag.value}
            style={{
              animation: `blinker 3s linear infinite, float 4s ease-in-out infinite`,
              animationDelay: `${Math.random() * 2}s, ${Math.random() * 2}s`,
              fontSize: `${size}px`,
              margin: '3px',
              padding: '3px',
              display: 'inline-block',
              cursor: 'pointer',
              color: color,
              fontWeight: size > 20 ? 'bold' : 'normal',
              opacity: 0.7 + (size / maxSize) * 0.3,
              transition: 'all 0.3s ease',
              textShadow: size > 25 ? '0 1px 2px rgba(0,0,0,0.1)' : 'none',
            }}
            onMouseEnter={(e) => {
              e.currentTarget.style.transform = 'scale(1.15)'
              e.currentTarget.style.opacity = '1'
              e.currentTarget.style.zIndex = '10'
            }}
            onMouseLeave={(e) => {
              e.currentTarget.style.transform = 'scale(1)'
              e.currentTarget.style.opacity = String(0.7 + (size / maxSize) * 0.3)
              e.currentTarget.style.zIndex = 'auto'
            }}
          >
            {_tag.value}
          </span>
        )}
      />
      <style jsx>{`
        .word-cloud-container {
          width: 100%;
          height: 200px;
          display: flex;
          align-items: center;
          justify-content: center;
          overflow: hidden;
          position: relative;
        }
        
        .word-cloud {
          text-align: center;
          line-height: 1.5;
          width: 100%;
          height: 100%;
        }
        
        @keyframes blinker {
          0% { opacity: 0.7; }
          50% { opacity: 1; }
          100% { opacity: 0.7; }
        }
        
        @keyframes float {
          0%, 100% { transform: translateY(0px); }
          50% { transform: translateY(-2px); }
        }
      `}</style>
    </div>
  )
} 