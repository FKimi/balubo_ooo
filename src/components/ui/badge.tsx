import * as React from "react"

interface BadgeProps extends React.HTMLAttributes<HTMLDivElement> {
  variant?: 'default' | 'secondary' | 'outline' | 'destructive'
}

const Badge = React.forwardRef<HTMLDivElement, BadgeProps>(
  ({ className = '', variant = 'default', ...props }, ref) => {
    const baseClasses = "inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors"
    
    let variantClasses = ""
    switch (variant) {
      case 'secondary':
        variantClasses = "border-gray-200 bg-gray-100 text-gray-800"
        break
      case 'outline':
        variantClasses = "border-gray-300 text-gray-700"
        break
      case 'destructive':
        variantClasses = "border-red-500 bg-red-500 text-white"
        break
      default:
        variantClasses = "border-blue-500 bg-blue-500 text-white"
    }

    return (
      <div
        ref={ref}
        className={`${baseClasses} ${variantClasses} ${className}`}
        {...props}
      />
    )
  }
)
Badge.displayName = "Badge"

export { Badge }
export type { BadgeProps } 