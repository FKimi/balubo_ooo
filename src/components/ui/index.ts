export * from "./badge";
export * from "./button";
export * from "./card";
export * from "./input";
export * from "./label";
export * from "./progress";
export * from "./textarea";
export * from "./TabNavigation";
export * from "./skeleton";
export * from "./WordCloud";
