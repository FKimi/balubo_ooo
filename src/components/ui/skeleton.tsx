import React from "react";

interface SkeletonProps {
  className?: string;
}

/**
 * 汎用スケルトンプレースホルダー
 * Usage:
 * <Skeleton className="w-full h-4" />
 */
export const Skeleton: React.FC<SkeletonProps> = ({ className = "" }) => {
  return (
    <div
      className={`animate-pulse rounded bg-gray-200 dark:bg-gray-700 ${className}`}
      aria-hidden="true"
    />
  );
};
