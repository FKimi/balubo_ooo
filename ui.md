はい、承知いたしました。
UI/UXデザインのユーザビリティチェックリストを、コピーしてそのまま`.md`ファイルとして利用できるMarkdown形式のコードブロックで提供します。

```markdown
# UI/UXデザイン ユーザビリティチェックリスト

この記事は、i3DESIGN Designers様のnote記事「実案件から学んだ、本当に役立つUIデザインの法則50」を基に、UIデザインの重要な法則をまとめたものです。ヤコブ・ニールセンが提唱した「10のヒューリスティクス」に沿って分類されています。

## 1. システムステータスの可視化 (Visibility of system status)
> ユーザーに対して、システムが今どういう状況なのかを適切なフィードバックを通じて常に知らせる必要があります。

### 1-1. 入力項目が多いときはステップを分ける
フォームの入力項目が多い場合、項目をグルーピングして画面を分割する。ステッパーを設置し、現在の進捗状況を明示することで、ユーザーのモチベーションを維持する。

### 1-2. モードを意識させる
編集モードなど、特定の操作が制限される状態では、ユーザーに「モード中」であることを明確に伝える。例えば、編集中であることを示すバーを表示したり、タブバーのような離脱動線を非表示にしたりする。

### 1-3. 画面の変化をアニメーションで表す
画面が一瞬で切り替わると、ユーザーは状態の変化を認識しにくい。アニメーションを用いることで、どのような過程で変化したのかが分かりやすくなり、システムの構造を直感的に理解できる。

### 1-4. オーバーレイする時は背景を透過する
モーダルウィンドウなどを表示する際は、背景を透過させる。これにより、ユーザーは新しいページに遷移したのではなく、元の画面の上に情報が表示されていると認識でき、混乱を防げる。

### 1-5. 絞り込み中の状態を明示する
検索結果画面などで絞り込みが行われている場合、その状態を明確に示す。「絞り込み条件の数」「ボタンの色の変化」「条件のチップ表示」などでユーザーに伝える。

### 1-6. エンプティステートを用意する
データやコンテンツがまだ何もない「空の状態（エンプティステート）」では、ユーザーが何をすべきか分かるように案内を表示する。画面の目的と、コンテンツを追加する方法をナビゲートする。

### 1-7. 状態が変化するときはアクションの近くで行う
ダウンロードの開始や完了など、状態が変化する際のフィードバックは、アクションを起こしたボタンなどの近くに表示する。ユーザーがフィードバックに気づきやすくなる。

### 1-8. サイト・アプリ上の現在地を把握できるようにする
ユーザーが迷わないように、タイトルやパンくずリストを設置し、常に自分がどこにいるのかを把握できるようにする。

### 1-9. フリップフロップ問題を避ける
ON/OFFを切り替えるボタンのラベルが、現在の状態か、押した後の状態か分かりにくい問題を避ける。チェックボックスや色、アイコンの追加などで、現在の状態を明確に示す。

---

## 2. システムと現実世界の一致 (Match between system and the real world)
> ユーザーにとって馴染みのある言葉や概念を使用し、情報を自然で論理的な順序で表示する必要があります。

### 2-1. ユーザーが分かる言葉を使う
「認証に失敗しました」のような専門用語ではなく、「パスワードは半角英数字8文字以上で入力してください」のように、ユーザーが理解でき、次にとるべき行動が分かる言葉を使う。

### 2-2. ボタン/リンクの文言は「押すと何が起きるのか」を明示する
「進む」「OK」ではなく、「入力内容を確認する」「送信する」のように、ボタンを押すことで何が起こるのかを具体的に記述する。

### 2-3. ボタンの文言を動詞的表現にする
モーダルダイアログのボタンは、「OK」や「はい」ではなく、「保存する」「削除する」のように、実行されるアクションを表す動詞を使う。

### 2-4. 日付と分かる表記にする
「-」（ハイフン）や「.」（ピリオド）ではなく、「/」（スラッシュ）を使うなど、一般的に日付と認識されやすい記号を用いる。一覧表示ではゼロパディングで桁を揃える。

---

## 3. ユーザーのコントロールと自由度 (User control and freedom)
> ユーザーが誤った操作をしても、簡単に元に戻したり、操作を中断したりできる「非常口」を明確に示す必要があります。

### 3-1. 常にユーザーがコントロールできるようにする
間違えたり、気が変わったりしたときに、操作を簡単に修正・やり直しできるようにし、ユーザーが常にシステムをコントロールできる状態を保つ。

### 3-2. カルーセルはドットインジケーターだけでなくボタンをつける
ドットインジケーターは現在位置を示すのには有効だが、クリックしにくい。コンテンツを切り替えるための「<」「>」などのボタンも併せて設置する。

### 3-3. スクロールの先を示唆する
横スクロールなどでコンテンツが続く場合、コンテンツの一部を画面の端から少しだけ見せることで、スクロールできることをユーザーに示唆する。

### 3-4. カード/リストはクリックできることを示唆する
カードやリストがクリック可能な場合、シャドウをつけたり、矢印アイコンを加えたりして、遷移先があることを視覚的に伝える。

---

## 4. 一貫性と標準化 (Consistency and standards)
> 異なる単語や状況、アクションが同じ意味なのか、ユーザーを迷わせてはいけません。業界やプラットフォームの慣例に従うべきです。

### 4-1. 遷移後のページのタイトルを揃える
「マイページ」ボタンを押したら、遷移先のページのタイトルも「マイページ」に揃える。ボタンの文言とページのタイトルを一致させることで、ユーザーの混乱を防ぐ。

### 4-2. タブだとわかる見た目にする
タブUIは、選択中のタブが明確で、コンテンツエリアと一体化しているように見せる。他のUI（例: Chips）と混同されない、タブとして一般的に認識されるデザインにする。

### 4-3. 機能が違うものを同じ見た目にしない
見た目が同じ要素は、同じ挙動をするとユーザーは期待する。機能が異なる要素（例: select要素とアコーディオン）は、見た目に明確な差をつける。

### 4-4. 機能と色を揃える
色の持つ一般的なイメージ（赤は危険・エラー、緑は安全・正常など）と、機能の意味を一致させる。「削除」ボタンには赤を使い、危険性を直感的に伝える。

### 4-5. リンクとボタンを使い分ける
**リンク**は「画面遷移」など、ユーザーが対象に影響を与えないアクションに使う。**ボタン**は「データの操作」など、ユーザーがシステムに影響を与えるアクションに使う。両者を明確に区別する。

### 4-6. テキストリンクであることを下線と色で明示する
テキストリンクは、他のテキストと区別できるように、色を変え、下線を引くのが基本。これにより、クリックできることが一目で分かる。

### 4-7. 外部リンクや新規タブを開くリンクにはアイコンをつける
ユーザーの意図しない挙動（外部サイトへの遷移や新規タブの表示）を防ぐため、該当するリンクにはアイコンを付け、通常の内部リンクと区別する。

---

## 5. エラー防止 (Error prevention)
> エラーメッセージも重要ですが、そもそもエラーが起こらないようにデザインすることが最も重要です。

### 5-1. リンクアクションは誤タップしないよう十分な間隔を保つ
小さなボタンやリンクが密集していると、誤タップの原因になる。要素のサイズに合わせて、十分な間隔を確保する。

### 5-2. 取り返しがつかないアクションは遠くに置く
「削除」のような取り返しのつかないアクションのボタンは、他の頻繁に使うボタンから離れた位置に配置し、誤って押してしまうのを防ぐ。

### 5-3. タップ/クリック可能なエリアを広く取る
ボタンやアイコンの見た目のサイズよりも、実際に反応するクリックエリアを少し広めに設定する。これにより、ユーザーのストレスを軽減する。

### 5-4. 条件は先に提示する
パスワードの文字数制限や、アップロードファイルのサイズ制限など、システム側の要求する条件は、ユーザーがアクションを起こす前に提示する。

### 5-5. スクロール時に重要な情報が隠れないようにする
テーブル（表）などをスクロールする際、見出し行や列を固定表示にする。今見ているデータが何の項目なのかを常に確認できるようにする。

### 5-6. リアルタイムバリデーションで条件の達成状況を明示する
パスワード設定などで複数の条件がある場合、入力中にどの条件を満たしているかをリアルタイムで表示する。エラーを未然に防ぎ、ユーザーの入力を補助する。

### 5-7. 手動入力を避ける
入力可能な値に制限がある場合（個数、項目など）、キーボードによる手動入力を避け、ステッパーやドロップダウン、スライダーなどを用いて選択させることでエラーを防ぐ。

---

## 6. 想起よりも認識 (Recognition rather than recall)
> ユーザーに情報を記憶させるのではなく、選択肢を提示することで、記憶への負荷を最小限に抑えます。

### 6-1. 入力候補を表示する
検索窓などでテキスト入力中に、入力候補（サジェスト）を表示する。入力の手間を省き、誤字脱字を防ぐ。

### 6-2. プレースホルダーで記入例を明示する
入力フォームには、何を入力すべきかのヒントとなる記入例をプレースホルダーで示す。ただし、プレースホルダーは入力時に消えるため、ラベルとしても併用しない。

### 6-3. 適切なフォーム幅にする
郵便番号や電話番号のように、入力文字数がある程度決まっているフォームは、その長さに合わせた幅にする。フォーム幅が入力内容のヒントになる。

---

## 7. 使用の柔軟性と効率性 (Flexibility and efficiency of use)
> 初心者だけでなく、熟練したユーザーも効率的に操作できるように、ショートカットなどの機能を提供します。

### 7-1. スクロール負荷を少なくする
FAQページなど、長いコンテンツが複数ある場合は、アコーディオンUIを用いて初期状態では内容を閉じておく。ユーザーが必要な情報だけを展開できるようにし、スクロールの負担を減らす。

### 7-2. 価値のあるデフォルト値にする
設定項目や選択肢において、多くのユーザーが選ぶであろう値をデフォルト（初期値）に設定する。生年月日の西暦など、ユーザーの入力負荷を軽減できる。

### 7-3. 自動入力できるものは入力させない
郵便番号から住所を自動入力するなど、他の情報から導き出せる項目は、ユーザーに手動で入力させず、システム側で自動入力する。

### 7-4. 選択肢が3個以下の時には、セレクト要素ではなくラジオボタンを使う
選択肢が少ない場合、クリックして選択肢を展開する手間がかかるセレクトボックスよりも、一目で全ての選択肢が見えるラジオボタンの方が効率的。

### 7-5. スイッチとチェックボックスを使い分ける
**スイッチ**は、操作が即座にシステムに反映される設定に使う（例: 通知のON/OFF）。**チェックボックス**は、複数の項目を選択し、「保存」ボタンなどで一括して適用する場合に使う。

---

## 8. 美的でミニマルなデザイン (Aesthetic and minimalist design)
> インターフェースには、不要な情報を含めない。余分な情報は重要な情報の視認性を低下させます。

### 8-1. 可読性の高いテキストデザインにする
行間は文字サイズの50%〜100%程度、1行の文字数は25〜40文字程度を目安にするなど、文章が読みやすいように行間や行長を調整する。

### 8-2. 錯視を考慮して視覚調整を行う
アイコンなどを中央に配置する際、数値上の中心ではなく、人間の目から見て中心に見える「視覚的な中心」に調整する。

### 8-3. シンプルにする
使用する色やフォントの種類を絞り、デザインをシンプルに保つ。要素を強調しすぎると、本当に伝えたい情報の優先度が下がってしまう。

### 8-4. むやみに情報を隠さない
見た目をスッキリさせるために、マウスホバーしないと表示されない情報や機能を作るのは避ける。重要な機能は常に表示し、全てのユーザーが平等にアクセスできるようにする。

### 8-5. ボタンの優先度を考慮して表現を変える
複数のボタンが並ぶ場合、最も重要なアクション（例: 保存）のボタンは目立つようにし、優先度の低いアクション（例: キャンセル）は控えめなデザインにする。

### 8-6. フォームでの入力項目はグループごとにまとめる
関連する入力項目（例: 氏名とフリガナ、住所情報など）は、線や余白でグループ化する。認知負荷が下がり、ユーザーが入力しやすくなる。

### 8-7. 色だけに頼らない
情報の伝達を色だけに依存しない。色覚特性を持つユーザーもいるため、形、アイコン、テキスト、下線などを併用し、色がなくても情報が伝わるようにする。

### 8-8. コントラスト比をクリアしたテキストカラーを使う
背景色とテキスト色のコントラスト比は、WCAGの基準である「4.5:1」以上を確保する。ユーザーがストレスなくテキストを読めるようにする。

### 8-9. 情報のかたまりの数を7つ以下にする
人間が短期記憶で一度に処理できる情報量は「7±2個」と言われている。ナビゲーションメニューなどの項目は、7つ以下に絞るか、カテゴリ分けをして整理する。

---

## 9. エラー時のユーザーによる認識・診断・回復の補助 (Help users recognize, diagnose, and recover from errors)
> エラーメッセージは分かりやすい言葉で表現し、問題点を正確に示し、解決策を提案する必要があります。

### 9-1. エラー内容をリアルタイムに明示する
フォーム入力エラーは、送信ボタンを押した後ではなく、入力フィールドのすぐ側にリアルタイムで表示する。ユーザーはどこを修正すればよいかすぐに分かり、負担が少ない。

---

## 10. ヘルプとドキュメント (Help and documentation)
> 直感的に使えるのが一番ですが、必要に応じて、ユーザーがタスクを完了するためのヘルプやドキュメントを提供する必要があります。

### 10-1. アイコンだけに頼らない
アイコンは直感的だが、意味が伝わらないとユーザーを混乱させる。テキストラベルと併用するか、アイコン単体で使う場合はツールチップで説明を補うなど、意味が確実に伝わるようにする。
```
