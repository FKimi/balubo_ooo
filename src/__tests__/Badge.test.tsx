import { render } from '@testing-library/react'
import { Badge } from '@/components/ui/badge'

describe('<Badge />', () => {
  it('renders correctly with default variant', () => {
    const { container } = render(<Badge>Default</Badge>)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders correctly with secondary variant', () => {
    const { container } = render(<Badge variant="secondary">Secondary</Badge>)
    expect(container.firstChild).toMatchSnapshot()
  })
})
