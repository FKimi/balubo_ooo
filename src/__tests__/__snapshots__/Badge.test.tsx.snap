// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Badge /> renders correctly with default variant 1`] = `
<div
  class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors border-blue-500 bg-blue-500 text-white "
>
  Default
</div>
`;

exports[`<Badge /> renders correctly with secondary variant 1`] = `
<div
  class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors border-gray-200 bg-gray-100 text-gray-800 "
>
  Secondary
</div>
`;
